mobility/choreographer/src/validator.cc

```diff
@@ -22,17 +22,150 @@
+            std::max(zone.min.y, zone.max.y), std::max(zone.min.z, zone.max.z);
+        Vec3f tmp = Vec3f::Zero();
+        for (int i = 0; i < 3; i++) {
+          int j = (i + 1) % 3;
+          int k = (i + 2) % 3;
+          for (auto zx = zmin(j); zx <= zmax(j); zx += map_res_) {
+            for (auto zy = zmin(k); zy <= zmax(k); zy += map_res_) {
+              tmp(j) = zx;
+              tmp(k) = zy;
+              tmp(i) = zmin(i) - map_res_ * 1.001;
+              map[jps_map_util_->getIndex(jps_map_util_->floatToInt(tmp))] =
+                  val_occ_;
+              tmp(i) = zmax(i) + map_res_ * 1.001;
+              map[jps_map_util_->getIndex(jps_map_util_->floatToInt(tmp))] =
+                  val_occ_;
+            }
+          }
+        }
+      }
+    }
+    // 2) We set the interior of the keepin zones to free, this corrects the
+    // case where keepin zones are connected and a contourn was put between them
+    for (auto &zone : zones_.zones) {
+      if (zone.type == ff_msgs::Zone::KEEPIN) {
+        zmin << std::min(zone.min.x, zone.max.x),
+            std::min(zone.min.y, zone.max.y), std::min(zone.min.z, zone.max.z);
+        zmax << std::max(zone.min.x, zone.max.x),
+            std::max(zone.min.y, zone.max.y), std::max(zone.min.z, zone.max.z);
+        // add points on surface
+        Vec3f tmp = Vec3f::Zero();
+        for (auto zx = zmin(0); zx <= zmax(0); zx += map_res_) {
+          for (auto zy = zmin(1); zy <= zmax(1); zy += map_res_) {
+            for (auto zz = zmin(2); zz <= zmax(2); zz += map_res_) {
+              tmp(0) = zx;
+              tmp(1) = zy;
+              tmp(2) = zz;
+              map[jps_map_util_->getIndex(jps_map_util_->floatToInt(tmp))] =
+                  val_free_;
+            }
+          }
+        }
+      }
+    }
+
+    // 4) Add keepout zones
+    for (auto &zone : zones_.zones) {
+      if (zone.type == ff_msgs::Zone::KEEPOUT) {
+        zmin << std::min(zone.min.x, zone.max.x),
+            std::min(zone.min.y, zone.max.y), std::min(zone.min.z, zone.max.z);
+        zmax << std::max(zone.min.x, zone.max.x),
+            std::max(zone.min.y, zone.max.y), std::max(zone.min.z, zone.max.z);
+        // add points on surface
+        Vec3f tmp = Vec3f::Zero();
+        for (int i = 0; i < 3; i++) {
+          int j = (i + 1) % 3;
+          int k = (i + 2) % 3;
+          for (auto zx = zmin(j); zx <= zmax(j); zx += map_res_) {
+            for (auto zy = zmin(k); zy <= zmax(k); zy += map_res_) {
+              tmp(j) = zx;
+              tmp(k) = zy;
+              tmp(i) = zmin(i);
+              map[jps_map_util_->getIndex(jps_map_util_->floatToInt(tmp))] =
+                  val_occ_;
+              tmp(i) = zmax(i);
+              map[jps_map_util_->getIndex(jps_map_util_->floatToInt(tmp))] =
+                  val_occ_;
+            }
+          }
+        }
+      }
+    }
+
+    // set voxel map using keepin/keepout zones
```

Comment:
If you can, might want to split this into multiple functions for the different steps